readthedocs/core/unresolver.py

```diff
@@ -0,0 +1,68 @@
+import logging
+from urllib.parse import urlparse
+from collections import namedtuple
+
+from django.urls import resolve as url_resolve
+from django.test.client import RequestFactory
+
+from readthedocs.core.utils.extend import SettingsOverrideObject
+from readthedocs.proxito.middleware import map_host_to_project_slug
+from readthedocs.proxito.views.utils import _get_project_data_from_request
+from readthedocs.proxito.views.mixins import ServeDocsMixin
+
+log = logging.getLogger(__name__)
+
+UnresolvedObject = namedtuple(
+    'Unresolved', 'project, language_slug, version_slug, filename, fragment')
+
+
+class UnresolverBase:
+
+    def unresolve(self, uri):
+        """
+        Turn a URL into the component parts that our views would use to process them.
+
+        This is useful for lots of places,
+        like where we want to figure out exactly what file a URL maps to.
+        """
+        parsed = urlparse(uri)
+        domain = parsed.netloc.split(':', 1)[0]
+        path = parsed.path
+
+        # TODO: Make this not depend on the request object,
+        # but instead move all this logic here working on strings.
+        request = RequestFactory().get(path=path, HTTP_HOST=domain)
+        project_slug = request.host_project_slug = map_host_to_project_slug(request)
+
+        # Handle returning a response
+        if hasattr(project_slug, 'status_code'):
+            return UnresolvedObject(None, None, None, None, None)
```

Comment:
Why not just return `None` instead?

Reply:
It will be easier to perform checks by checking `if parts` than `if parts.<somefield>` --looks more intuitive to me, since `<somefield>` could be `None` and be a valid object as well (`fragment`, for example)

Reply:
Makes sense 👍